0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+
+Here are some examples of common patterns expressed in terms of array broadcasting, and their accompanying visual representation in the figure that follows:
+
+```python
+# Broadcast single observable
+parameter_values = np.array(1.0)
+parameter_values.shape == ()
+observables = ObservablesArray(["III", "XXX", Pauli("YYY"), "ZZZ", Pauli("XYZ")])
+observables.shape == (5,)
+>> task_result.shape == (5,)
+
+# Zip
+parameter_values = BindingsArray(np.random.uniform(size=(5,)))
+parameter_values.shape == (5,)
+observables = ObservablesArray([Pauli("III"), Pauli("XXX"), Pauli("YYY"), Pauli("ZZZ"), Pauli("XYZ")])
+observables.shape == (5,)
+>> task_result.shape == (5,)
+
+# Outer/Product
+parameter_values = BindingsArray(np.random.uniform(size=(1, 6)))
+parameter_values.shape == (1, 6)
+observables = ObservablesArray([[Pauli("III")], [Pauli("XXX")], [Pauli("YYY")], [Pauli("ZZZ")]])
+observables.shape == (4, 1)
+>> task_result.shape == (4,6)
+
+# Standard nd generalization
+parameter_values = BindingsArray(np.random.uniform(size=(3, 6)))
+parameter_values.shape == (3, 6)
+observables = ObservablesArray([
+    [[Pauli(...)], [Pauli(...)], [Pauli(...)]]],
+    [[Pauli(...)], [Pauli(...)], [Pauli(...)]]],
+])
+observables.shape == (2, 3, 1)
+>> task_result.shape == (3, 6, 2)
+```
+
+<img src="./0015-estimator-interface/broadcasting.svg">
+
+### Type Coercion Strategy <a name="type-coercion-strategy"></a>
+
+To minimize the number of container types that an every-day user will need to interact with, and to make the transition more seamless, we propose that several container types be associated with a `TypeLike` pattern and a static method
+
+```python
+def coerce(argument: TypeLike) -> Type
+```
+
+that is invoked inside of the `run()` method.
+
+For example,
+
+```python
+BindingsArrayLike = Union[BindingsArray, ArrayLike, Mapping[Parameter, float]]
+```
+
+and
+
+```python
+@staticmethod
+def coerce(bindings_array: BindingsArrayLike) -> BindingsArray:
+    if isinstance(bindings_array, (list, ArrayLike)):
+        bindings_array = BindingsArray(bindings_array)
+    elif isinstance(bindings_array, Mapping):
+        bindings_array = BindingsArray(kwargs=bindings_array)
+
+    return bindings_array
+```
+
+In particular, we propose this kind of Coercion for the types:
+* `EstimatorTask`
+* `BindingsArray`
+* `ObservablesArray`
+
+### TaskResults <a name="TaskResults"></a>
+
+The results from each `EstimatorTask` will be array valued, and each `EstimatorTask` in the same job may have a different shape.
+Consider an `EstimatorTask` with shape `<20, 30>`, where the shape has come from the broadcasting rules discussed elsewhere. 
+This will result in a 20×30 array of real estimates.
+Moreover, we will want to return an array of standard deviations of the same shape.
+This would result in a bundle of broadcastable arrays:
+
+```python
+TaskResult({“evs”: <20, 30>, “stds”: <20, 30>}, metadata)
+```
```

Comment:
Just writing down something I mentioned in the meeting, for posterity:

There's questions about whether standard deviation or error should be returned, but I think there's general consensus that what's returned there should have some relation to the "error bars" we'd draw on a graph, so they're representative of some sort of confidence interval.

When error mitigation is in the picture, this confidence interval is not necessarily symmetric. If nothing else, the raw estimation of the result of a single shot isn't naturally normally distributed but binomially, but in general, mitigation strategies aren't going to have their best estimates of the confidence interval be symmetric.  Giving a single number here is never going to be sufficient to describe the whole underlying multivariate distribution, so the only useful thing to return is some approximate quantity that can be given a clear, universal meaning.

The std/variance/std.error based on a univariate normal approximation is one choice there.  Another would be to define that the return value has shape `<20, 30, 2>`, with the indices being the 5-95% confidence bounds on the value (or whatever arbitrary choice of percentiles), and say that an implementer of `Estimator` is free to return more detailed statistical information in the `metadata` in a form of their choosing.  There's still no concept of the results because drawn from a multivariate distribution, so it's just choosing different places to make the approximation in.  I think a confidence interval on the marginal of each variable might be more easily defined/implemented for a general interface than "the std.err on the 'closest' normal approximation to the marginal for some implementer-defined definition of 'closest'" (or whatever).

It could also be an option to supply a couple of instance methods on `TaskResult` that take whatever uncertainty information we end up requiring of implementers, and calculate an approximation to the stddev etc.